plugins/splunk-on-call/src/api/client.ts

```diff
@@ -0,0 +1,204 @@
+  PatchIncidentRequest,
+} from './types';
+
+export class UnauthorizedError extends Error {}
+
+export const splunkOnCallApiRef = createApiRef<SplunkOnCallApi>({
+  id: 'plugin.splunk-on-call.api',
+  description: 'Used to fetch data from Splunk On-Call API',
+});
+
+export class SplunkOnCallClient implements SplunkOnCallApi {
+  static fromConfig(configApi: ConfigApi, discoveryApi: DiscoveryApi) {
+    const usernameFromConfig: string | null =
+      configApi.getOptionalString('splunkoncall.username') || null;
+    return new SplunkOnCallClient({
+      username: usernameFromConfig,
+      discoveryApi,
+    });
+  }
+  constructor(private readonly config: ClientApiConfig) {}
+
+  async getIncidents(): Promise<Incident[]> {
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/incidents`;
+
+    const { incidents } = await this.getByUrl<IncidentsResponse>(url);
+
+    return incidents;
+  }
+
+  async getOnCallUsers(): Promise<OnCall[]> {
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/oncall/current`;
+    const { teamsOnCall } = await this.getByUrl<OnCallsResponse>(url);
+
+    return teamsOnCall;
+  }
+
+  async getTeams(): Promise<Team[]> {
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/team`;
+    const teams = await this.getByUrl<Team[]>(url);
+
+    return teams;
+  }
+
+  async acknowledgeIncident({
+    incidentNames,
+  }: PatchIncidentRequest): Promise<Response> {
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        userName: this.config.username,
+        incidentNames,
+      }),
+    };
+
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/incidents/ack`;
+
+    return this.request(url, options);
+  }
+
+  async resolveIncident({
+    incidentNames,
+  }: PatchIncidentRequest): Promise<Response> {
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        userName: this.config.username,
+        incidentNames,
+      }),
+    };
+
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/incidents/resolve`;
+
+    return this.request(url, options);
+  }
+
+  async getUsers(): Promise<User[]> {
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v2/user`;
+    const { users } = await this.getByUrl<ListUserResponse>(url);
+
+    return users;
+  }
+
+  async getEscalationPolicies(): Promise<EscalationPolicyInfo[]> {
+    const url = `${await this.config.discoveryApi.getBaseUrl(
+      'proxy',
+    )}/splunk-on-call/v1/policies`;
```

Comment:
Notice two different API version tags in here perhaps?  I know absolutely nothing about it, but just pointing it out in case they do represent two different vers?  (`/splunk-on-call/v2/user` vs `/splunk-on-call/v1/policies`)

I was going to suggest the static `/splunk-on-call/v2` might be a good thing to externalize into a const, but the version change threw me 😁

Reply:
Yes I noticed that.
I wanted to use only version 2 of the API but some routes are only available in v1 (for incidents for example).
You can look at the [API documentation](https://portal.victorops.com/api-docs/).

Reply:
Note I saw this on their web site:

> Note that the REST endpoint is different than the general API and is the preferred method to create incidents.

REST Endpoint: https://help.victorops.com/knowledge-base/rest-endpoint-integration-guide/ (that's where the above quote comes from)
API: https://help.victorops.com/knowledge-base/api/ (which then links to the link you had)

I claim no advanced knowledge of this interface, just reading the manual 😁

Reply:
We could use this API and it would make more sense as it is the recommended method.

I found it a bit complicated for the end user, to have to generate an ID API and a KEY API + a REST Endpoint for the creation-resolution-acknowledgement of the incident. That's why I decided to stay with the API.

But I think it's the best approach since it allows us to better configure our incidents.

Do I make the changes in this PR or do we open an issue to discuss this in more detail and iterate later on this version?

Reply:
I'd say if you know this works, likely just stick with it, and maybe either file an issue or put on your TODO to refactor using the vendor's recommended go forward? The community might get good use out of all of the work you've already put forth.

Reply:
Okay, I'll file an issue and I'll assign it to myself 👍